File: apps/agent/src/practices/dayClock.ts
/**
 * Shared time helper for practice directives that gate on the user's
 * local-clock window or stamp idempotency by local date. `en-CA` reliably
 * returns YYYY-MM-DD in the requested timezone; `sv-SE` returns 24-hour
 * HH without locale surprises.
 *
 * Originally inlined in eveningGratitude; extracted when day_planning
 * needed the same primitive.
 */

export function localDateAndHour(now: Date, tz: string | null): { date: string; hour: number } {
  const tzOpt = tz ?? 'UTC';
  const date = new Intl.DateTimeFormat('en-CA', {
    timeZone: tzOpt,
    year: 'numeric',
    month: '2-digit',
    day: '2-digit',
  }).format(now);
  const hourStr = new Intl.DateTimeFormat('sv-SE', {
    timeZone: tzOpt,
    hour: '2-digit',
    hour12: false,
  }).format(now);
  return { date, hour: Number.parseInt(hourStr, 10) };
}

File: apps/agent_py/src/lifecoach_agent/practices/day_clock.py
"""Local-clock helpers shared by practice directives that gate on a
specific time-of-day window or stamp idempotency by coaching date.

Mirrors `apps/agent/src/practices/dayClock.ts`. The TS version uses
`Intl.DateTimeFormat` with locales chosen for predictable output
(en-CA gives YYYY-MM-DD, sv-SE gives 0–23 hours). In Python we use
zoneinfo + strftime which gives the same shape directly.
"""

from __future__ import annotations

from datetime import datetime

from lifecoach_agent.coaching_day import coaching_date_and_hour


def local_date_and_hour(now: datetime, tz: str | None) -> tuple[str, int]:
    """Return `(coaching YYYY-MM-DD, hour-0-23)` in `tz`.

    `tz=None` falls back to UTC. `now` is treated as a timezone-aware
    `datetime` — naive datetimes are interpreted as UTC for safety.
    """
    return coaching_date_and_hour(now, tz)
